### PR TITLE
feat: MLSys 2026 DAG Scheduler — Track A (Rust) + Track B (Gemini Agent)

### LINKED ISSUE #1: feat: topological sort for DAG linearization
## F-14: Topological Sort
Given any valid DAG, return operations in a valid linearized execution order.
### Acceptance Criteria
- [ ] Handles all 5 benchmark DAG structures
- [ ] Tested against examples 1-5 from PROBLEM.md
- [ ] Returns deterministic ordering
### Effort: 1h | Dependencies: none

### LINKED ISSUE #2: feat: problem JSON parser
## F-01: Problem JSON Parser
Parse all benchmark JSON files into internal Problem data structures.
### Acceptance Criteria
- [ ] Reads all 5 benchmark files + example_problem.json without error
- [ ] Reconstructs Problem struct with correct tensor/op counts and hardware params
- [ ] Maps to Python dataclasses matching C++ structs from mlsys.h
### Effort: 2h | Dependencies: none

### LINKED ISSUE #3: feat: latency model matching C++ Evaluate()
## F-02: Latency Model
Implement latency calculation matching the C++ Evaluate() function exactly.
### Acceptance Criteria
- [ ] Passes all 5 worked-example test cases from PROBLEM.md with latencies matching to 0.1 precision
- [ ] Handles compute-bound, memory-bound, and tiled (multi-step) cases correctly
### Effort: 4h | Dependencies: F-01

### LINKED ISSUE #4: feat: working-set calculator with OOM guard
## F-03: Working-Set Calculator
Calculate memory usage for a subgraph and detect OOM violations.
### Acceptance Criteria
- [ ] Given a subgraph + resident tensor set + granularity, returns working-set size
- [ ] Raises OOM flag if working set > fast_memory_capacity
- [ ] Verified against all 5 PROBLEM.md examples
### Effort: 3h | Dependencies: F-01, F-02

### LINKED ISSUE #5: feat: regression tests for all 5 PROBLEM.md examples
## F-13: Regression Tests
Test suite covering all 5 worked examples from PROBLEM.md.
### Acceptance Criteria
- [ ] Covers all 5 examples with Strategies A/B/C where given
- [ ] All tests pass before optimizer code is written
- [ ] Validates latency values to within 0.1 precision
### Effort: 3h | Dependencies: F-02, F-03

### LINKED ISSUE #6: feat: solution JSON serializer
## F-11: Solution JSON Serializer
Write well-formed solution JSON matching the output schema.
### Acceptance Criteria
- [ ] Writes valid JSON matching output schema
- [ ] Round-trips through JSON validator
- [ ] null traversal_orders serialize correctly
### Effort: 2h | Dependencies: none

### LINKED ISSUE #7: feat: baseline scheduler (1 op per subgraph)
## F-04: Baseline Scheduler
Produce a trivially correct schedule: one subgraph per operation, native granularity.
### Acceptance Criteria
- [ ] One valid subgraph per operation
- [ ] Uses native granularity [128, 128, K_full]
- [ ] tensors_to_retain = [] for all subgraphs
- [ ] Latency values match model; no OOM on any benchmark
### Effort: 2h | Dependencies: F-01, F-02, F-03, F-11, F-14

### LINKED ISSUE #8: feat: CLI benchmark runner
## F-12: Benchmark Runner
CLI tool to evaluate solutions against problems.
### Acceptance Criteria
- [ ] Accepts --problem FILE --solution FILE arguments
- [ ] Calls evaluate logic, prints total latency and pass/fail
### Effort: 3h | Dependencies: F-01, F-11

### LINKED ISSUE #9: feat: op grouping / chain fusion optimizer
## F-05: Op Grouping / Chain Fusion
Greedy bottom-up fusion: group adjacent ops if merged working set fits in fast memory.
### Acceptance Criteria
- [ ] Groups adjacent ops in topological order when memory allows
- [ ] Intermediate tensors become ephemeral (zero cost, zero capacity)
- [ ] Verify latency improves vs. baseline on all 5 benchmarks
### Effort: 6h | Dependencies: F-14, F-03, F-02

### LINKED ISSUE #10: feat: tensor retention across subgraph boundaries
## F-07: Tensor Retention
Decide which output tensors to retain in fast memory across subgraph boundaries.
### Acceptance Criteria
- [ ] After each subgraph, determine which outputs consumed by next subgraph
- [ ] Retain if sufficient residual capacity exists
- [ ] Verify improvement on Example 3C pattern
### Effort: 4h | Dependencies: F-05, F-03

### LINKED ISSUE #11: feat: split-K for memory-constrained MatMuls
## F-08: Split-K
For MatMul subgraphs where full-k working set exceeds capacity, search for largest k divisor that fits.
### Acceptance Criteria
- [ ] Search k divisors that fit within fast memory
- [ ] Model accumulator as resident across k-steps
- [ ] Verify Example 5B latency matches
### Effort: 5h | Dependencies: F-05, F-03, F-02

### LINKED ISSUE #12: feat: granularity search for optimal [w,h,k] per subgraph
## F-06: Granularity Search
For each subgraph, search candidate [w, h] values to minimize latency while satisfying memory constraints.
### Acceptance Criteria
- [ ] Try candidate [w, h] values (powers of 2 up to tensor dimensions)
- [ ] Select granularity that minimizes subgraph latency within OOM constraint
- [ ] Verify Example 1C pattern
### Effort: 8h | Dependencies: F-05, F-03, F-02

### PR DESCRIPTION
**Pull request by melroyanthony**
## Summary

Full implementation of the MLSys 2026 contest DAG scheduler with two tracks:

- **Track A (Rust)**: Compiled binary `mlsys` with 9-stage optimizer pipeline
- **Track B (Python)**: Gemini-powered agent with local evaluator fallback

Closes #1, Closes #2, Closes #3, Closes #4, Closes #5, Closes #6, Closes #7, Closes #8, Closes #9, Closes #10, Closes #11, Closes #12

## Architecture

```mermaid
graph LR
    A[Problem JSON] --> B[Parser]
    B --> C[DAG Analysis]
    C --> D[Baseline]
    D --> E[Chain Fusion]
    E --> F[Retention 1]
    F --> G[Split-K]
    G --> H[Granularity Search]
    H --> I[Retention 2]
    I --> J[Emergency OOM Fix]
    J --> K[Latency Recalc]
    K --> L[Traversal Opt]
    L --> M[Solution JSON]
```

## Optimizer Pipeline (9 stages)

1. **Baseline**: 1 op per subgraph, native granularity — guarantees valid output
2. **Chain Fusion**: Greedy merge of adjacent ops when working set fits and boundary output dims are consistent
3. **Retention (pass 1)**: Keep boundary tensors resident across subgraph boundaries
4. **Split-K**: Reduce k dimension (using min K_full) for memory-constrained MatMuls
5. **Granularity Search**: Exhaustive (w, h, k) search per subgraph for minimum latency
6. **Retention (pass 2)**: Re-evaluate after granularity changes
7. **Emergency OOM Fix**: Reduce granularity for any remaining OOM subgraphs
8. **Final Latency Recalculation**: Recompute all subgraph latencies
9. **Traversal Optimization**: Snake/zig-zag tile ordering for MatMul data reuse

## CLI Interface

```bash
# Track A — Solve
./mlsys <input.json> <output.json>

# Track A — Evaluate existing solution
./mlsys evaluate --problem <input.json> --solution <solution.json>

# Track B — Python agent
GOOGLE_API_KEY=<key> uv run python agent.py <input.json> <output.json>
```

## Test Results

- **Track A (Rust)**: 15 tests passing (9 worked examples + 6 edge cases)
- **Track B (Python)**: Local evaluator + optimizer validated
- **E2E**: Both tracks validated against all 5 benchmarks

## Benchmark Results (Track A — Rust)

| Benchmark | Ops | Tensors | Fast Mem | Bandwidth | Latency |
|-----------|-----|---------|----------|-----------|---------|
| mlsys-2026-1  | 5   | 9   | 60,000   | 20  | 27,443    |
| mlsys-2026-5  | 19  | 29  | 30,000   | 15  | 27,856    |
| mlsys-2026-9  | 32  | 49  | 250,000  | 25  | 110,100   |
| mlsys-2026-13 | 63  | 100 | 600,000  | 50  | 191,693   |
| mlsys-2026-17 | 103 | 160 | 500,000  | 100 | 23,650    |

All benchmarks complete in under 1 second.

## SDLC Pipeline Stages

| Stage | Status | Judge Score |
|-------|--------|-------------|
| 0: Setup | Done | — |
| 1: Requirements (RICE/MoSCoW) | Done | 91/100 |
| 2: Architecture & Design | Done | 88/100 |
| 2.5: Issues + Branch | Done | 12 issues |
| 3: Implementation (Rust + Python) | Done | 82/100 |
| 4: Testing & Validation | Done | 15 tests, 0 bugs |
| 5: Docs, CI/CD, Finalization | Done | — |

## Key Fixes During Review

- Per-op K_full scaling: each MatMul uses its own K_full (not boundary MatMul's)
- Split-K uses min K_full across ops (safe for mixed-K subgraphs)
- Fusion validates boundary output dimension consistency
- Parser validates op_type, MatMul arity, tensor bounds
- Snake traversal optimization (Stage 9)
- Evaluate subcommand for solution validation
- All planning docs updated to reflect Rust implementation

## Test plan

- [x] `cargo test` — 15 unit tests covering all 5 PROBLEM.md examples
- [x] All 5 benchmarks produce valid JSON with non-negative latencies
- [x] Evaluate subcommand validates solutions (PASS/FAIL + latency)
- [x] Track B runs in baseline mode without API key
- [x] E2E script validates both tracks against all benchmarks

### CODE REVIEW

File: solution/scripts/test-e2e.sh [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+# Tests both Track A (Rust) and Track B (Python) against all 5 benchmarks.
+# Run from the project root: bash solution/scripts/test-e2e.sh
+
+set -uo pipefail
+
+PROJECT_ROOT="/Users/melroyanthony/Desktop/area-51/projects/MLSys"
+RUST_DIR="$PROJECT_ROOT/solution/backend/rust"
+AGENT_DIR="$PROJECT_ROOT/solution/agent"
+BENCH_DIR="$PROJECT_ROOT/problem/benchmarks"
+TMP_DIR="/tmp/mlsys-e2e-$$"
```
[the remark this answers is not in the document]
Comment (melroyanthony) [reply solution/scripts/test-e2e.sh]: Fixed: PROJECT_ROOT now derived dynamically via dirname.

File: solution/README.md [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+**Entry point:** `solution/agent/`
+
+---
+
+## Quick Start
+
+### Prerequisites
+
+- Rust toolchain 1.80+ (`rustup`)
+- Python 3.11+ and `uv` (`pip install uv` or `curl -LsSf https://astral.sh/uv/install.sh | sh`)
```
[the remark this answers is not in the document]
Comment (melroyanthony) [reply solution/README.md]: Fixed: README now says Python 3.12+.

File: solution/backend/rust/src/optimizer/fusion.rs [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+                    .iter()
+                    .chain(groups[i + 1].iter())
+                    .copied()
+                    .collect();
+
+                let retained_before: HashSet<usize> = HashSet::new();
+
+                if find_feasible_granularity(&merged, &retained_before, problem, dag).is_some() {
+                    new_groups.push(merged);
+                    i += 2;
```
[the remark this answers is not in the document]
Comment (melroyanthony) [reply solution/backend/rust/src/optimizer/fusion.rs]: Fixed: added boundary output dimension consistency check before merging.

File: solution/backend/rust/src/dag.rs [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+        let num_tensors = problem.tensors.len();
+
+        let mut tensor_producer: Vec<Option<usize>> = vec![None; num_tensors];
+        let mut tensor_consumers: Vec<Vec<usize>> = vec![vec![]; num_tensors];
+
+        for (op_idx, op) in problem.ops.iter().enumerate() {
+            for &out_t in &op.outputs {
+                tensor_producer[out_t] = Some(op_idx);
+            }
+            for &in_t in &op.inputs {
```
[the remark this answers is not in the document]
Comment (melroyanthony) [reply solution/backend/rust/src/dag.rs]: Fixed: added bounds checking on tensor indices with descriptive errors.

File: solution/backend/rust/src/optimizer/pipeline.rs [hunk 5]
```diff
@@ -0,0 +1,153 @@
+/// Pipeline: orchestrates all optimizer stages in sequence.
+///
+/// Stages:
+/// 1. Start with baseline (one op per subgraph, native granularity)
+/// 2. Greedy fusion (merge adjacent ops)
+/// 3. Split-K (handle OOM from large k)
+/// 4. Granularity search (find best w, h, k per subgraph)
```
[the remark this answers is not in the document]
Comment (melroyanthony) [reply solution/backend/rust/src/optimizer/pipeline.rs]: Fixed: updated header comment to reflect actual 9-stage pipeline (retention x2, emergency OOM, traversal).

File: solution/scripts/test-e2e.sh [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+
+# ---------------------------------------------------------------------------
+# Track A: Build Rust binary
+# ---------------------------------------------------------------------------
+echo "--- Track A: Building Rust binary ---"
+cd "$RUST_DIR"
+if cargo build --release 2>/dev/null; then
+    pass "Track A: Rust build"
+else
+    fail "Track A: Rust build failed"
```
[the remark this answers is not in the document]
Comment (melroyanthony) [reply solution/scripts/test-e2e.sh]: Fixed: Track A benchmarks are now skipped on build failure. Removed stderr suppression from cargo build.

File: solution/requirements/rice-scores.md [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+| Rank | Feature | Score | Notes |
+|------|---------|-------|-------|
+| 1 | F-14: Topological sort | 30.0 | Lowest effort, unblocks everything |
+| 2 | F-01: Problem JSON parser | 15.0 | Gate to all other work |
+| 2 | F-11: Solution JSON serializer | 15.0 | Gate to evaluation |
+| 3 | F-02: Latency model (roofline) | 7.5 | Core correctness |
+| 4 | F-13: Regression tests | 6.7 | Validate latency model early |
+| 5 | F-03: Working-set calculator | 10.0 | Prevents invalid schedules |
+| 6 | F-05: Op grouping / chain fusion | 5.0 | Primary latency reduction lever |
```
[the remark this answers is not in the document]
Comment (melroyanthony) [reply solution/requirements/rice-scores.md]: Fixed: RICE table reordered — F-03 (10.0) now correctly ranked above F-02 (7.5).

File: solution/scripts/test-e2e.sh [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+
+# No duplicate ops
+op_counts = {}
+for sg in subgraphs:
+    for op in sg:
+        op_counts[op] = op_counts.get(op, 0) + 1
+dupes = {k: v for k, v in op_counts.items() if v > 1}
+if dupes:
+    print(f'Duplicate ops: {dupes}', file=sys.stderr)
+    sys.exit(6)
```
[the remark this answers is not in the document]
Comment (melroyanthony) [reply solution/scripts/test-e2e.sh]: Fixed: relaxed duplicate op check to allow recomputation (ops may appear in multiple subgraphs per requirements).
